Register: error: use of deleted function

Trying to use `CtrlMeas::Mode1 | CtrlMeas::Mode0` with `register.hpp` in the plain blink example:

	#include <xpcc/architecture/platform.hpp>
	#include <xpcc/architecture/interface/register.hpp>

	using namespace Board;

	namespace xpcc
	{

	struct bme280
	{
	protected:

		enum class
		CtrlMeas : uint8_t
		{
			OSRS_T2 = xpcc::Bit7,
			OSRS_T1 = xpcc::Bit6,
			OSRS_T0 = xpcc::Bit5,
			OSRS_P2 = xpcc::Bit4,
			OSRS_P1 = xpcc::Bit3,
			OSRS_P0 = xpcc::Bit2,
			Mode1   = xpcc::Bit1,
			Mode0   = xpcc::Bit0,
		};
		XPCC_FLAGS8(CtrlMeas);

		enum class
		Mode : uint8_t
		{
			Sleep = 0,
			Forced = CtrlMeas::Mode0,
			Normal = CtrlMeas::Mode1 | CtrlMeas::Mode0,
		};

	};

	} // xpcc namespace

	// ----------------------------------------------------------------------------
	int
	main()
	{
		while (true)
			{}
		return 0;
	}

With `gcc version 5.2.1 20151202`

	arm-none-eabi-g++ -v
	Using built-in specs.
	COLLECT_GCC=arm-none-eabi-g++
	COLLECT_LTO_WRAPPER=/opt/homebrew-cask/Caskroom/gcc-arm-embedded/5_2-2015q4,20151219/gcc-arm-none-eabi-5_2-2015q4/bin/../lib/gcc/arm-none-eabi/5.2.1/lto-wrapper
	Target: arm-none-eabi
	Configured with: /Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/src/gcc/configure --target=arm-none-eabi --prefix=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native --libexecdir=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native/lib --infodir=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native/share/doc/gcc-arm-none-eabi/info --mandir=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native/share/doc/gcc-arm-none-eabi/man --htmldir=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native/share/doc/gcc-arm-none-eabi/html --pdfdir=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native/share/doc/gcc-arm-none-eabi/pdf --enable-languages=c,c++ --enable-plugins --disable-decimal-float --disable-libffi --disable-libgomp --disable-libmudflap --disable-libquadmath --disable-libssp --disable-libstdcxx-pch --disable-nls --disable-shared --disable-threads --disable-tls --with-gnu-as --with-gnu-ld --with-newlib --with-headers=yes --with-python-dir=share/gcc-arm-none-eabi --with-sysroot=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/install-native/arm-none-eabi --build=x86_64-apple-darwin10 --host=x86_64-apple-darwin10 --with-gmp=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/build-native/host-libs/usr --with-mpfr=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/build-native/host-libs/usr --with-mpc=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/build-native/host-libs/usr --with-isl=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/build-native/host-libs/usr --with-cloog=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/build-native/host-libs/usr --with-libelf=/Users/build/workspace/GCC-5-0-build_build-toolchain-mac/gcc-arm-none-eabi-5_2-2015q4-20151219/build-native/host-libs/usr --with-host-libstdcxx='-static-libgcc -Wl,-lstdc++ -lm' --with-pkgversion='GNU Tools for ARM Embedded Processors' --with-multilib-list=armv6-m,armv7-m,armv7e-m,armv7-r,armv8-m.base,armv8-m.main
	Thread model: single
	gcc version 5.2.1 20151202 (release) [ARM/embedded-5-branch revision 231848] (GNU Tools for ARM Embedded Processors)

yields to

	Compiling C++: /Users/usr/Dev/xpcc/build/stm32f4_discovery/blink/main.o
		main.cpp:32:40: error: use of deleted function 'constexpr xpcc::Register<T>::operator unsigned char() const [with T = unsigned char]'
		   Normal = CtrlMeas::Mode1 | CtrlMeas::Mode0,
		                                        ^
		In file included from /Users/usr/Dev/xpcc/src/xpcc/architecture/interface.hpp:107:0,
		                 from /Users/usr/Dev/xpcc/build/stm32f4_discovery/blink/libxpcc/src/xpcc/architecture/../../../generated_platform/driver/can/generic/can_bit_timings.hpp:13,
		                 from /Users/usr/Dev/xpcc/build/stm32f4_discovery/blink/libxpcc/src/xpcc/architecture/../../../generated_platform/drivers.hpp:36,
		                 from /Users/usr/Dev/xpcc/build/stm32f4_discovery/blink/libxpcc/src/xpcc/architecture/platform.hpp:19,
		                 from main.cpp:1:
		/Users/usr/Dev/xpcc/src/xpcc/architecture/interface/register.hpp:264:12: note: declared here
		  constexpr operator unsigned char() const = delete;
		            ^
		scons: *** [/Users/usr/Dev/xpcc/build/stm32f4_discovery/blink/main.o] Error 1
		scons: building terminated because of errors.

IMHO the usage is the same as in this [example] (https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/interface/register.hpp#L126) but is not covered by the [unittest] (https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/interface/test/register_test.hpp#L70). Shall the example work?

When leaving out the `| CtrlMeas::Mode0` part the compiler is happy.

It shall be noted that `XPCC_FLAGS8` is only usable inside a  `struct` or `class`, so putting a `struct` around your register definitions is necessary.

When I change

```
v1 = Test3::Bit | Config2(Config::Three) | Config0(Config::Two);
```

to

```
v1 = Config2(Config::Three) | Config0(Config::Two);
```

With `g++-5`

```
g++-5 -o /Users/usr/Dev/xpcc/build/unittest_hosted/xpcc/architecture/interface/test/register_test.o -c -std=c++11 -Woverloaded-virtual -funsigned-char -Wall -Wextra -Wundef -ggdb -DBASENAME=register_test.cpp -DXPCC__CLOCK_TESTMODE=1 -I. -I/Users/usr/Dev/xpcc/ext -I/Users/usr/Dev/xpcc/build/unittest_hosted/libxpcc/src xpcc/architecture/interface/test/register_test.cpp
    xpcc/architecture/interface/test/register_test.cpp: In member function 'void RegisterTest::testConfigurations()':
    xpcc/architecture/interface/test/register_test.cpp:271:30: error: ambiguous overload for 'operator|' (operand types are 'xpcc::testing::Config2 {aka xpcc::Configuration<xpcc::Flags<xpcc::testing::Test3, unsigned char>, xpcc::testing::Config, 3u, 2u>}' and 'xpcc::testing::Config0 {aka xpcc::Configuration<xpcc::Flags<xpcc::testing::Test3, unsigned char>, xpcc::testing::Config, 3u>}')
      v1 = Config2(Config::Three) | Config0(Config::Two);
```

Same with `g++-6`

```
g++-6 -o /Users/usr/Dev/xpcc/build/unittest_hosted/xpcc/architecture/interface/test/register_test.o -c -std=c++11 -Woverloaded-virtual -funsigned-char -Wall -Wextra -Wundef -ggdb -DBASENAME=register_test.cpp -DXPCC__CLOCK_TESTMODE=1 -I. -I/Users/usr/Dev/xpcc/ext -I/Users/usr/Dev/xpcc/build/unittest_hosted/libxpcc/src xpcc/architecture/interface/test/register_test.cpp
xpcc/architecture/interface/test/register_test.cpp: In member function 'void RegisterTest::testConfigurations()':
xpcc/architecture/interface/test/register_test.cpp:271:30: error: ambiguous overload for 'operator|' (operand types are 'xpcc::testing::Config2 {aka xpcc::Configuration<xpcc::Flags<xpcc::testing::Test3, unsigned char>, xpcc::testing::Config, 3u, 2u>}' and 'xpcc::testing::Config0 {aka xpcc::Configuration<xpcc::Flags<xpcc::testing::Test3, unsigned char>, xpcc::testing::Config, 3u>}')
  v1 = Config2(Config::Three) | Config0(Config::Two);
       ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
```

`Apple LLVM version 7.3.0 (clang-703.0.31)`

```
g++ -o /Users/usr/Dev/xpcc/build/unittest_hosted/xpcc/architecture/interface/test/register_test.o -c -std=c++11 -Woverloaded-virtual -funsigned-char -Wall -Wextra -Wundef -ggdb -DBASENAME=register_test.cpp -DXPCC__CLOCK_TESTMODE=1 -I. -I/Users/usr/Dev/xpcc/ext -I/Users/usr/Dev/xpcc/build/unittest_hosted/libxpcc/src xpcc/architecture/interface/test/register_test.cpp
xpcc/architecture/interface/test/register_test.cpp:271:30: error: use of overloaded operator '|' is ambiguous (with operand types 'Config2' (aka 'Configuration<xpcc::Flags<xpcc::testing::Test3,
      unsigned char>, xpcc::testing::Config, Bit1 | Bit0, 2>') and 'Config0' (aka 'Configuration<xpcc::Flags<xpcc::testing::Test3, unsigned char>, xpcc::testing::Config, Bit1 | Bit0>'))
        v1 = Config2(Config::Three) | Config0(Config::Two);
             ~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~
```

--- a/src/xpcc/architecture/interface/register.hpp
+++ b/src/xpcc/architecture/interface/register.hpp
@@ -288,7 +288,152 @@ using Register32 = Register<uint32_t>;	///< Register class with 32-bit wide unde
  * Class for operating on a register.
  *
  * This class extends the Register base class by adding constructor
- * and bitwise operation overloading and convenience functions between
+ * and bitwise operation overloading between
+ * a strongly-typed enum and itself.
+ * It is the common base class for Flags, Configurations and Values!
+ *
+ * @tparam	Enum	a strongly-typed enum containing the bit masks
+ * @tparam	T		the underlying integer type to be for internal representation
+ *
+ * @see Flags
+ *
+ * @ingroup	register
+ * @author	Niklas Hauser
+ */
+template < typename Enum, typename T >
+struct FlagsOperators : public ::xpcc::Register<T>
+{
+	/// @cond
+	typedef T UnderlyingType;
+	using Register<T>::value;
+	/// @endcond
+
+	typedef Enum EnumType;	///< The enum type
+
+	// MARK: Constructors and assignments
+
+	/// @{
+
+	/// default constructor initializes `value` to zero
+	constexpr FlagsOperators() {}
+
+	/// constructs itself out of a strongly-typed enum
+	constexpr FlagsOperators(Enum flag) :
+		Register<T>(T(flag)) {}
+
+	/// When explicitly called, this class constructs itself out of the underlying type
+	explicit constexpr FlagsOperators(UnderlyingType flag) :
+		Register<T>(flag) {}
+
+	/// When explicitly called, this class constructs itself out of the superclass
+	explicit constexpr FlagsOperators(Register<T> const &o) :
+		Register<T>(o.value) {}
+
+	constexpr FlagsOperators(FlagsOperators const &o) = default;	///< Copy constructor
+	constexpr FlagsOperators(FlagsOperators &&o) = default;		///< Move constructor
+	/// @}
+
+	/// @{
+	FlagsOperators& operator=(FlagsOperators const &o) = default;	///< Copy assignment
+	FlagsOperators& operator=(FlagsOperators &&o) = default;		///< Move assignment
+	/// @}
+
+	static constexpr FlagsOperators mask()	/// returns the mask of the integer type
+	{ return FlagsOperators(T(-1)); }
+
+	// flags binary operations
+	/// bitwise negation
+	constexpr FlagsOperators operator~() const { return FlagsOperators(~value); }
+
+	/// @{
+
+	/// bitwise **AND** with multiple bits
+	constexpr FlagsOperators operator&(FlagsOperators const &o) const { return FlagsOperators(value & o.value); }
+	/// bitwise **OR** with multiple bits
+	constexpr FlagsOperators operator|(FlagsOperators const &o) const { return FlagsOperators(value | o.value); }
+	/// bitwise **XOR** with multiple bits
+	constexpr FlagsOperators operator^(FlagsOperators const &o) const { return FlagsOperators(value ^ o.value); }
+	/// @}
+
+	/// @{
+
+	/// bitwise **AND** with multiple bits
+	FlagsOperators& operator&=(FlagsOperators const &o) {  return (value &= o.value, *this); }
+	/// bitwise **OR** with multiple bits
+	FlagsOperators& operator|=(FlagsOperators const &o) { return (value |= o.value, *this); }
+	/// bitwise **XOR** with multiple bits
+	FlagsOperators& operator^=(FlagsOperators const &o) { return (value ^= o.value, *this); }
+	/// @}
+
+	/// @{
+
+	/// bitwise **AND** with a single bit
+	FlagsOperators& operator&=(Enum const &flag) { return operator&=(FlagsOperators(flag)); }
+	/// bitwise **OR** with a single bit
+	FlagsOperators& operator|=(Enum const &flag) { return operator|=(FlagsOperators(flag)); }
+	/// bitwise **XOR** with a single bit
+	FlagsOperators& operator^=(Enum const &flag) { return operator^=(FlagsOperators(flag)); }
+	/// @}
+
+	/// @{
+
+	/// bitwise **AND** with a single bit
+	friend constexpr FlagsOperators operator&(Enum const &a, FlagsOperators const &b) { return b & a; }
+	/// bitwise **OR** with a single bit
+	friend constexpr FlagsOperators operator|(Enum const &a, FlagsOperators const &b) { return b | a; }
+	/// bitwise **XOR** with a single bit
+	friend constexpr FlagsOperators operator^(Enum const &a, FlagsOperators const &b) { return b ^ a; }
+	/// @}
+
+#ifndef __DOXYGEN__
+	constexpr FlagsOperators operator&(Enum const &flag) const { return operator&(FlagsOperators(flag)); }
+	constexpr FlagsOperators operator|(Enum const &flag) const { return operator|(FlagsOperators(flag)); }
+	constexpr FlagsOperators operator^(Enum const &flag) const { return operator^(FlagsOperators(flag)); }
+
+	constexpr bool operator==(Enum const &flag) const { return value == T(flag); }
+	constexpr bool operator!=(Enum const &flag) const { return value != T(flag); }
+
+	friend constexpr bool operator==(Enum const &a, FlagsOperators const &b) { return b == a; }
+	friend constexpr bool operator!=(Enum const &a, FlagsOperators const &b) { return b != a; }
+
+	constexpr bool operator==(FlagsOperators const &o) const { return value == o.value; }
+	constexpr bool operator!=(FlagsOperators const &o) const { return value != o.value; }
+#else
+	/// @{
+
+	/// bitwise **AND** with a single bit
+	friend constexpr FlagsOperators operator&(FlagsOperators const &a, Enum const &b);
+	/// bitwise **OR** with a single bit
+	friend constexpr FlagsOperators operator|(FlagsOperators const &a, Enum const &b);
+	/// bitwise **XOR** with a single bit
+	friend constexpr FlagsOperators operator^(FlagsOperators const &a, Enum const &b);
+	/// @}
+
+	/// @{
+
+	/// bitwise **AND** with a single bit
+	friend constexpr FlagsOperators operator&(Enum const &a, Enum const &b);
+	/// bitwise **OR** with a single bit
+	friend constexpr FlagsOperators operator|(Enum const &a, Enum const &b);
+	/// bitwise **XOR** with a single bit
+	friend constexpr FlagsOperators operator^(Enum const &a, Enum const &b);
+	/// @}
+
+	/// Compares if two flags are equal
+	/// @{
+	friend constexpr bool operator==(FlagsOperators const &a, Enum const &b);
+	friend constexpr bool operator!=(FlagsOperators const &a, Enum const &b);
+	friend constexpr bool operator==(Enum const &a, FlagsOperators const &b);
+	friend constexpr bool operator!=(Enum const &a, FlagsOperators const &b);
+	/// @}
+#endif
+} ATTRIBUTE_PACKED;
+
+/**
+ * Class for operating on a register.
+ *
+ * This class extends the FlagsOperators base class by adding constructor
+ * overloading and convenience functions between
  * a strongly-typed enum and itself.
  *
  * This class makes heavy use of `constexpr`, so that as many values
@@ -321,11 +466,11 @@ using Register32 = Register<uint32_t>;	///< Register class with 32-bit wide unde
  * @author	Niklas Hauser
  */
 template < typename Enum, typename T >
-struct Flags : public ::xpcc::Register<T>
+struct Flags : public ::xpcc::FlagsOperators<Enum, T>
 {
 	/// @cond
 	typedef T UnderlyingType;
-	using Register<T>::value;
+	using FlagsOperators<Enum, T>::value;
 	/// @endcond
 
 	typedef Enum EnumType;	///< The enum type
@@ -339,15 +484,18 @@ struct Flags : public ::xpcc::Register<T>
 
 	/// constructs itself out of a strongly-typed enum
 	constexpr Flags(Enum flag) :
-		Register<T>(T(flag)) {}
+		FlagsOperators<Enum, T>(flag) {}
+
+	constexpr Flags(FlagsOperators<Enum, T> const &flag) :
+		FlagsOperators<Enum, T>(flag) {}
 
 	/// When explicitly called, this class constructs itself out of the underlying type
 	explicit constexpr Flags(UnderlyingType flag) :
-		Register<T>(flag) {}
+		FlagsOperators<Enum, T>(flag) {}
 
 	/// When explicitly called, this class constructs itself out of the superclass
 	explicit constexpr Flags(Register<T> const &o) :
-		Register<T>(o.value) {}
+		FlagsOperators<Enum, T>(o) {}
 
 	constexpr Flags(Flags const &o) = default;	///< Copy constructor
 	constexpr Flags(Flags &&o) = default;		///< Move constructor
@@ -357,9 +505,6 @@ struct Flags : public ::xpcc::Register<T>
 	Flags& operator=(Flags const &o) = default;	///< Copy assignment
 	Flags& operator=(Flags &&o) = default;		///< Move assignment
 	/// @}
-
-	static constexpr Flags mask()	/// returns the mask of the integer type
-	{ return Flags(T(-1)); }
 
 	/// @{
 	// MARK: convenience methods
@@ -423,93 +568,6 @@ struct Flags : public ::xpcc::Register<T>
 	constexpr bool none(Flags const &o) const
 	{ return (*this & o).value == 0; }
 	/// @}
-
-	// flags binary operations
-	/// bitwise negation
-	constexpr Flags operator~() const { return Flags(~value); }
-
-	/// @{
-
-	/// bitwise **AND** with multiple bits
-	constexpr Flags operator&(Flags const &o) const { return Flags(value & o.value); }
-	/// bitwise **OR** with multiple bits
-	constexpr Flags operator|(Flags const &o) const { return Flags(value | o.value); }
-	/// bitwise **XOR** with multiple bits
-	constexpr Flags operator^(Flags const &o) const { return Flags(value ^ o.value); }
-	/// @}
-
-	/// @{
-
-	/// bitwise **AND** with multiple bits
-	Flags& operator&=(Flags const &o) {  return (value &= o.value, *this); }
-	/// bitwise **OR** with multiple bits
-	Flags& operator|=(Flags const &o) { return (value |= o.value, *this); }
-	/// bitwise **XOR** with multiple bits
-	Flags& operator^=(Flags const &o) { return (value ^= o.value, *this); }
-	/// @}
-
-	/// @{
-
-	/// bitwise **AND** with a single bit
-	Flags& operator&=(Enum const &flag) { return operator&=(Flags(flag)); }
-	/// bitwise **OR** with a single bit
-	Flags& operator|=(Enum const &flag) { return operator|=(Flags(flag)); }
-	/// bitwise **XOR** with a single bit
-	Flags& operator^=(Enum const &flag) { return operator^=(Flags(flag)); }
-	/// @}
-
-	/// @{
-
-	/// bitwise **AND** with a single bit
-	friend constexpr Flags operator&(Enum const &a, Flags const &b) { return b & a; }
-	/// bitwise **OR** with a single bit
-	friend constexpr Flags operator|(Enum const &a, Flags const &b) { return b | a; }
-	/// bitwise **XOR** with a single bit
-	friend constexpr Flags operator^(Enum const &a, Flags const &b) { return b ^ a; }
-	/// @}
-
-#ifndef __DOXYGEN__
-	constexpr Flags operator&(Enum const &flag) const { return operator&(Flags(flag)); }
-	constexpr Flags operator|(Enum const &flag) const { return operator|(Flags(flag)); }
-	constexpr Flags operator^(Enum const &flag) const { return operator^(Flags(flag)); }
-
-	constexpr bool operator==(Enum const &flag) const { return value == T(flag); }
-	constexpr bool operator!=(Enum const &flag) const { return value != T(flag); }
-
-	friend constexpr bool operator==(Enum const &a, Flags const &b) { return b == a; }
-	friend constexpr bool operator!=(Enum const &a, Flags const &b) { return b != a; }
-
-	constexpr bool operator==(Flags const &o) const { return value == o.value; }
-	constexpr bool operator!=(Flags const &o) const { return value != o.value; }
-#else
-	/// @{
-
-	/// bitwise **AND** with a single bit
-	friend constexpr Flags operator&(Flags const &a, Enum const &b);
-	/// bitwise **OR** with a single bit
-	friend constexpr Flags operator|(Flags const &a, Enum const &b);
-	/// bitwise **XOR** with a single bit
-	friend constexpr Flags operator^(Flags const &a, Enum const &b);
-	/// @}
-
-	/// @{
-
-	/// bitwise **AND** with a single bit
-	friend constexpr Flags operator&(Enum const &a, Enum const &b);
-	/// bitwise **OR** with a single bit
-	friend constexpr Flags operator|(Enum const &a, Enum const &b);
-	/// bitwise **XOR** with a single bit
-	friend constexpr Flags operator^(Enum const &a, Enum const &b);
-	/// @}
-
-	/// Compares if two flags are equal
-	/// @{
-	friend constexpr bool operator==(Flags const &a, Enum const &b);
-	friend constexpr bool operator!=(Flags const &a, Enum const &b);
-	friend constexpr bool operator==(Enum const &a, Flags const &b);
-	friend constexpr bool operator!=(Enum const &a, Flags const &b);
-	/// @}
-#endif
 } ATTRIBUTE_PACKED;
 
 /// @ingroup	register
@@ -680,19 +738,20 @@ struct FlagsGroup<T> : public Register<typename T::UnderlyingType>
  * @author	Niklas Hauser
  */
 template < typename Parent, typename Enum, typename Parent::UnderlyingType Mask, typename Parent::UnderlyingType Position = 0 >
-struct Configuration : public ::xpcc::Register<typename Parent::UnderlyingType>
+struct Configuration : public ::xpcc::FlagsOperators<typename Parent::EnumType, typename Parent::UnderlyingType>
 {
 private:
 	typedef typename Parent::UnderlyingType UnderlyingType;
 	typedef UnderlyingType UType;	// lazy
+	typedef typename Parent::EnumType EType;	// lazy
 
 public:
 	/// explicit constructor for the underlying type
 	explicit constexpr Configuration(UnderlyingType config) :
-		Register<UType>(config << Position) {}
+		FlagsOperators<EType, UType>(config << Position) {}
 	/// explicit constructor for the enum type
 	explicit constexpr Configuration(Enum config) :
-		Register<UType>(UType(config) << Position) {}
+		FlagsOperators<EType, UType>(UType(config) << Position) {}
 
 	/// copy constructor
 	constexpr Configuration(Configuration const &o) = default;
@@ -720,7 +779,7 @@ public:
 
 	/// @cond
 	constexpr operator Parent() const
-	{	return Parent(Register<UType>::value); }
+	{	return Parent(FlagsOperators<EType, UType>::value); }
 	/// @endcond
 } ATTRIBUTE_PACKED;
 
@@ -746,15 +805,18 @@ public:
  * @author	Niklas Hauser
  */
 template < typename Parent, typename Parent::UnderlyingType Width, typename Parent::UnderlyingType Position = 0 >
-struct Value : public ::xpcc::Register<typename Parent::UnderlyingType>
+struct Value : public ::xpcc::FlagsOperators<typename Parent::EnumType, typename Parent::UnderlyingType>
 {
 private:
-	typedef typename Parent::UnderlyingType PType;
-	static constexpr PType Mask = ((1 << Width) - 1) << Position;
+	typedef typename Parent::UnderlyingType UnderlyingType;
+	typedef UnderlyingType UType;	// lazy
+	typedef typename Parent::EnumType EType;	// lazy
+
+	static constexpr UType Mask = ((1 << Width) - 1) << Position;
 public:
 	/// explicit constructor for the underlying type
-	explicit constexpr Value(typename Parent::UnderlyingType config) :
-		Register<PType>((config << Position) & Mask) {}
+	explicit constexpr Value(UnderlyingType config) :
+		FlagsOperators<EType, UType>((config << Position) & Mask) {}
 
 	/// copy constructor
 	constexpr Value(Value const &o) = default;
@@ -764,7 +826,7 @@ public:
 	/// @{
 
 	/// clears and sets a new value in a Flags register
-	static inline void set(Parent &parent, typename Parent::UnderlyingType config)
+	static inline void set(Parent &parent, UnderlyingType config)
 	{	parent.value = (parent.value & ~Mask) | ((config << Position) & Mask); }
 
 	/// clears the value in a Flags register
@@ -772,7 +834,7 @@ public:
 	{	parent.value &= ~Mask; }
 
 	/// returns the value from a Flags register
-	static constexpr PType get(Parent const &parent)
+	static constexpr UType get(Parent const &parent)
 	{	return (parent.value & Mask) >> Position; }
 	/// @}
 
@@ -782,7 +844,7 @@ public:
 
 	/// @cond
 	constexpr operator Parent() const
-	{	return Parent(Register<PType>::value); }
+	{	return Parent(FlagsOperators<EType, UType>::value); }
 	/// @endcond
 } ATTRIBUTE_PACKED;
 

--- a/src/xpcc/architecture/interface/test/register_test.cpp
+++ b/src/xpcc/architecture/interface/test/register_test.cpp
@@ -268,6 +268,12 @@ RegisterTest::testConfigurations()
 	v1 = Test3::Bit | Config0(Config::Two);
 	TEST_ASSERT_EQUALS(v1.value, (xpcc::Bit4 | 0b10));
 
+	v1 = Config0(Config::Two) | Test3::Bit;
+	TEST_ASSERT_EQUALS(v1.value, (xpcc::Bit4 | 0b10));
+
+	v1 = Config2(Config::Three) | Config0(Config::Two);
+	TEST_ASSERT_EQUALS(v1.value, ((0b11 << 2) | 0b10));
+
 	v1 = Test3::Bit | Config2(Config::Three) | Config0(Config::Two);
 	TEST_ASSERT_EQUALS(v1.value, (xpcc::Bit4 | (0b11 << 2) | 0b10));
 
@@ -295,6 +301,12 @@ void
 RegisterTest::testValue()
 {
 	Test3_t v1 = Test3::Bit | Config0(Config::Two) | Address(6);
+	TEST_ASSERT_EQUALS(v1.value, (xpcc::Bit4 | 0b10 | (6 << 5)));
+
+	v1 = Config0(Config::Two) | Address(6) | Test3::Bit;
+	TEST_ASSERT_EQUALS(v1.value, (xpcc::Bit4 | 0b10 | (6 << 5)));
+
+	v1 = Address(6) | Config0(Config::Two) | Test3::Bit;
 	TEST_ASSERT_EQUALS(v1.value, (xpcc::Bit4 | 0b10 | (6 << 5)));
 
 	Address::reset(v1);

